Add bind:true option to getCurrentContext

josieusa opened the pull request:
### Description
Workaround issue #17 with cujojs/when v3.7.7 and similar packages (mostly `Promise` libraries) that use queues for async-related stuff.
**User** code is cleaner and more elegant than initially proposed in #17, because he/she only needs to remember to call `LoopBackContext.getCurrentContext()` at the first line of his/her middlewares/operation hooks which use the current context feature (if it's a remote hook, the feature should not be needed anymore; if you use it anyway, use it preferably in `beforeRemote` rather than `afterRemote`).
P.S. : use it early in the middleware/hook chain for a given route.
**Library** code, instead, can be improved a lot; this is just my first attempt. Style needs to be checked too.
Also, I only patched the context functions `get` and `set`, because I don't use other functions. I didn't patch the other functions yet (~I deleted them, whoops 😄  This aspect needs to be fixed~).

Thank you!

#### Related issues

<!--
Please use the following link syntaxes:

- #49 (to reference issues in the current repository)
- strongloop/loopback#49 (to reference issues in another repository)
-->

- #17

### Checklist

<!--
Please mark your choice with an "x" (i.e. [x], see
https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments)
-->

- [x] New tests added to cover all changes (the new test passes!)
- [x] Documentation changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

bajtos commented on server/middleware/per-request.js:
```diff
@@ -31,6 +31,16 @@ function perRequestContextFactory(options) {
   var scope = options.name || name;
   var enableHttpContext = options.enableHttpContext || false;
   var ns = LoopBackContext.createContext(scope);
+  // Monkey-patch LoopBackContext.getCurrentContext in order to fix issue #17
```
I am wondering - why are we applying this workaround only in the middleware, instead of binding get/set methods directly inside the original `getCurrentContext`?

Also IIUC, this code is fixating the `context` object to a single value when a request reaches `per-request` middleware. How is this going to support the case when there are multiple requests handled in parallel?

Imagine the following handler chain (three different middleware handlers):

 1. per-context
 2. access-token retrieval, finishes in a later tick
 3. store access token in the context

Now consider two requests handled in the following sequence:

 1. Request1 invokes per-context, modifies global `LoopBackContext.getCurrentContext` to return CTX1
 2. Request1 starts access-token retrieval
 3. Request2 invokes per-context, modifies global `LoopBackContext.getCurrentContext` to return CTX2
 4. Request2 starts access-token retrieval
 5. Request1 access token is retrieved
 6. The access token is stored in CTX2. Ooops.

josieusa replied:
I tried to bind inside the original method, but for some reasons it didn't work.
It should work if I monkey-patch the method inside, say, a utility module. This will be my next attempt (less dirty than the first), unless I figure out why it doesn't work in the original method. Please wait for the commit.
I agree that it needs to be moved away from the middleware factory. In fact, if you add the middleware to the app, it works (even outside the middleware); but if you don't add the middleware to the app... whoops.

josieusa replied:
Regarding the concurrent requests, this is done *exactly* for them. Even if the code needs to be fixed, tests pass with the workaround (and fail if I comment out the workaround), and it already works in a test app with 10 concurrent `curl` requests.

I have a possible explanation. Are you sure that points 1 and 3 are correct? In these points, the function that gets invoked is this:
https://github.com/strongloop/loopback-context/blob/v1.0.0/server/middleware/per-request.js#L36
*This* function (the middleware) does *not* modify the global LoopBackContext.getCurrentContext (it's the outer function, i.e. the middleware factory, that does, and the latter gets called when *adding* the middleware to the app).

josieusa replied:
Fixed, now the workaround is outside the middleware [factory], and it still works (and this time, this is what I expected) in my test app with 10 concurrent `curl` requests.

And tests pass *only* with the workaround in place. They fail as soon as you comment the workaround out. 👍 

Also, I refactored the original method. I no longer monkey-patch it, which was not beautiful.

I bet it can be done without the proxy, but since I couldn't manage to do it, I'm fine with the proxy.

The proxy makes it clearer, I hope. I delegate the `get` and `set` methods of the cls namespace to a local object (`boundMethods`) which gets created every time you call `getCurrentContext()` (I used a new object so that the binding of `this` when you call `ctx.get` or `ctx.set` is predictable, and can't get mixed up between requests).

And at the same time, I re-bind the methods to the (globally visible) namespace; more precisely, I bind them to the local variable in the body of `createContext()`which holds a reference to the (globally visible) namespace.

bajtos commented on test/main.test.js:
```diff
@@ -129,4 +130,50 @@ describe('LoopBack Context', function() {
+    var timeout = 50;
+    // Concurrent execution number 1 of 2
+    var execution1 = new Promise(function execution1(outerResolve, reject) {
+      LoopBackContext.runInContext(function pushToContext1() {
```
AFAICT, these tests are not invoking `per-request` middleware, therefore they are not verifying the change proposed in that middleware :(

josieusa replied:
Whoops, you are right, they are not invoking the middleware (I don't need it in the test; I only need parallel execution). The tests pass anyway (and a test app with concurrent requests works), *after* adding the workaround. But there is one more reason, that is, the middleware factory was called before, somewhere... I'll need to fix that. Tests should pass for only one reason (i.e. the addition of the fix).

josieusa replied:
Problem solved without changing the tests, since the change is no longer in the middleware [factory] (I moved it inside the original method, instead of patching it in the middleware [factory])

bajtos commented on test/main.test.js:
```diff
@@ -129,4 +130,50 @@ describe('LoopBack Context', function() {
   });
+  it('doesn\'t mix up contexts if using concurrently then() from when 3.7.7',
+  function() {
+    expect(require('when/package.json').version).to.equal('3.7.7');
```
I am little bit worried about this long-distance coupling between the tests and package.json, and also what will happen when we will want to test two different versions of a same package.

Could we run `npm install when@3.7.7` from the test instead? To speed up local `npm test` and prevent repeated fetching of packages from the registry, we can use https://www.npmjs.com/package/strong-cached-install to cache the results - I have had good experience using this approach in other LoopBack projects. We will need to expose API for installing a single package, but that should be trivial - see the existing [installPackage](https://github.com/strongloop/strong-cached-install/blob/91e7e428150d32749bbb1e066f127f42935d3d70/index.js#L51) function.

josieusa replied:
WIP

josieusa replied:
PTAL
Solved in a different way (local modules and stub modules)
Inspiration from here:
Local modules http://stackoverflow.com/a/26028854
Stub modules https://github.com/npm/npm/issues/5499#issuecomment-71089080
Works for me using two different, conflicting versions of the same package.

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

josieusa committed:
Workaround issue #17 with cujojs/when

josieusa commented:
Sorry, this is a test comment to see if the pull requests gets re-opened and the disappeared commit re-appears...
**EDIT**
I deleted the old commits (instead of reverting them), because the code was wrong, then I pushed a new commit, where I address all the line notes (except for the addition of `strong-cached-install`, which isn't done yet).

josieusa committed:
Fix corner case of missing context

jainisenough commented:
This issue also comes when we use `loopback-connector-rest` (latest version, 2.1.0) & call any third part api.

josieusa commented:
@jainisenough Sorry, I made a mistake in issue #17. Fixed. The issue is not that the context gets *lost*, but rather *mixed up between concurrent HTTP requests*. Sorry if I confused you.

josieusa committed:
Use local stub module for multi-versions

josieusa committed:
Remove call to Function.prototype.bind()

josieusa commented:
I just wanted to write a reminder that this PR doesn't seem to fix strongloop/loopback#2519

josieusa commented:
I just wanted to write a reminder that this PR doesn't seem to fix strongloop/loopback#2519 (I suspect it's because this PR only works for user code, not 3rd-party code invoking "cls-unaware" libraries).

jainisenough commented:
@josieusa, No. this is context lost issue when we call third party api via loopback-context-rest. i'm aware about loopback-connector-mongodb@1.15.3 issue, so I'm using 1.17.0 right now.

josieusa commented:
@jainisenough the issue strongloop/loopback-context#17 is that the context gets mixed up between requests.

josieusa committed:
Add comment about methods not re-bound

bajtos commented on package.json:
```diff
@@ -26,13 +26,14 @@
   },
   "devDependencies": {
-    "async": "1.5.2",
+    "async-1.5.2": "file:./test/stub-modules/async-1.5.2",
```
Nice trick 👍

superkhau replied:
Will this work on windows?

bajtos replied:
Using local paths for dependencies is a regular npm feature described in npm docs, see https://docs.npmjs.com/files/package.json#local-paths

I would expect it to work on Windows too.

bajtos commented on server/current-context.js:
```diff
@@ -78,7 +78,28 @@ LoopBackContext.createContext = function(scopeName) {
     // Set up LoopBackContext.getCurrentContext()
     LoopBackContext.getCurrentContext = function() {
-      return ns && ns.active ? ns : null;
+      if (ns && ns.active) {
```
```js
if (!ns || !ns.active)
  return null;

// the rest of the code is indented one level less
```

bajtos commented on server/current-context.js:
```diff
@@ -78,7 +78,28 @@ LoopBackContext.createContext = function(scopeName) {
+            target[name];
+          },
+        };
+        return new Proxy(ns, handler);
```
I find using `Proxy` rather heavy weight. How about using regular prototypal inheritance, would it not work equally well?

```js
var boundContext = Object.create(ns);
boundContext.get = ns.bind(ns.get);
boundContext.set = ns.bind(ns.set);
return boundContext;
```

bajtos replied:
`Proxy` also seems to be not available in Node 4.x, see https://travis-ci.org/strongloop/loopback-context/jobs/189233795#L264

```
Unhandled error for request GET /TestModels/test:
ReferenceError: Proxy is not defined
    at Object.LoopBackContext.getCurrentContext (/home/travis/build/strongloop/loopback-context/server/current-context.js:99:20)
```

bajtos commented on test/main.test.js:
```diff
@@ -129,4 +129,49 @@ describe('LoopBack Context', function() {
+          setTimeout(resolve, timeout);
+        });
+        whenPromise.then(function pullFromContext2() {
+          var testValue = ctx && ctx.get('test-key', 'test-value-2');
```
Isn't the second argument of `get` ignored?

bajtos commented on test/main.test.js:
```diff
@@ -129,4 +129,49 @@ describe('LoopBack Context', function() {
+  function() {
+    var timeout = 50;
+    // Concurrent execution number 1 of 2
+    var execution1 = new Promise(function execution1(outerResolve, reject) {
```
Considering that the code building execution1 and execution2 differ only in one parameter (the value stored in the context), the code would be much easier to read if the creation of execution promise was refactored into a shared function.

```js
Promise.all([
  runTestWithValue('test-value-1'),
  runTestWithValue('test-value-2'),
]);

function runTestWithValue(value) {
  LoopBackContext.runInContext(...);
}
```

bajtos commented on test/main.test.js:
```diff
@@ -129,4 +129,49 @@ describe('LoopBack Context', function() {
+        expect(ctx).is.an('object');
+        ctx.set('test-key', 'test-value-1');
+        var whenPromise = whenV377.promise(function(resolve) {
+          setTimeout(resolve, timeout);
```
Can we simplify this code by using `promise.delay`? See https://github.com/cujojs/when/blob/master/docs/api.md#promisedelay

bajtos commented on test/main.test.js:
```diff
@@ -129,4 +129,49 @@ describe('LoopBack Context', function() {
+        }).then(function verify1(testValue) {
+          expect(testValue).to.equal('test-value-1');
+          outerResolve();
+        }).catch(function(error) {
```
`.catch(reject)`

bajtos commented on server/current-context.js:
```diff
@@ -78,7 +78,28 @@ LoopBackContext.createContext = function(scopeName) {
+      if (ns && ns.active) {
+        /**
+         * **NOTE**
+         * This only re-binds get and set methods, the most used.
```
I am worried about breaking applications that expects `getCurrentContext` to return a value that is not bound yet.

How about introducing a new API for getting the bound context? E.g. `loopbackContext.getBoundContext`? Alternatively, we can add an `options` object to `getCurrentContext`, e.g. `loopbackContext.getCurrentContext({ bind: true })`.

Thoughts?

bajtos reviewed (changes requested):
Thank you @josieusa for the update, the patch looks much better now! I still have few concerns to discuss and/or address, see below.

josieusa committed:
Remove Proxy, add bind option, style

josieusa commented:
@bajtos all done. I chose the `getCurrentContext({bind: true})` alternative. PTAL

josieusa committed:
Document bind option of getCurrentContext()

josieusa committed:
Edit docs of bind option of getCurrentContext()

josieusa commented:
I added a "failure test" to check that it **fails** as expected with the `bind` option disabled (which is the default). The test passes on Node 6 but not on Node 4, as I see from Travis CI. This is a non-issue; it means that this use case of `loopback-context` works *better* (and not worse!) on Node 4 than in Node 6. But should I check the Node version inside the test? (i.e. should I run the added failure test only on Node 6 where the test is known to fail?)
**EDIT** I ran the tests locally with Node 4.7.2, same version as Travis CI. The failure test passed, unlike Travis.
**EDIT 2** I re-ran the tests locally using all versions of Node, and the failure test fails about every 20-30 runs. Then you run it again and it passes. What do you think about it? Travis CI always fails on Node 4. The only substantial difference I found is that Travis CI is slower and the timeout is low (25 ms).

josieusa committed:
Edit docs of bind option of getCurrentContext()

bajtos commented:
>  I re-ran the tests locally using all versions of Node, and the failure test fails about every 20-30 runs. Then you run it again and it passes. What do you think about it? Travis CI always fails on Node 4. The only substantial difference I found is that Travis CI is slower and the timeout is low (25 ms).

A test must either always pass or always fail. Tests that sometimes fail and sometimes pass only learn people to ignore failing tests, which will eventually lead to the situation when a legitimate test failure was ignored.

So, I think you can either try to change the test to make it more robust, for example don't use `Promise.all` but start the second task from the next tick, or perhaps do something else to change the timing. If we are not able to find any reliable solution, then it's better to disable the test completely (possibly only on Node v4).

bajtos commented on README.md:
```diff
@@ -62,6 +62,36 @@ This approach should be compatible with all process managers, including
 `strong-pm`. However, we feel that relying on the order of `require` statements
 is error-prone.
 
+### Bind for concurrency
```
I think this section should go after `### Use the current context` just before ` ### Use current authenticated user in remote methods`.

bajtos commented on test/main.test.js:
```diff
@@ -129,4 +129,40 @@ describe('LoopBack Context', function() {
+          expect(ctx).is.an('object');
+          ctx.set('test-key', pushedValue);
+          var whenPromise = whenV377.promise(function(resolve) {
+            setTimeout(resolve, timeout);
```
Can we simplify this code by using `promise.delay`? See https://github.com/cujojs/when/blob/master/docs/api.md#promisedelay

josieusa replied:
I didn't miss this one, but for some reason, it didn't work. Now it works 🤣

bajtos commented on test/main.test.js:
```diff
@@ -129,4 +129,40 @@ describe('LoopBack Context', function() {
+  it('handles concurrent then() calls with when v3.7.7 promises & bind option',
+  function() {
+    var timeout = 25;
+    function runWithPushedValue(pushedValue, bind, expectToFail) {
```
Boolean args are evil. Please use an options argument instead: `{bind: true}`.

josieusa replied:
I missed this one, too

josieusa committed:
Move bind section in README, remove bad test

bajtos commented on test/main.test.js:
```diff
@@ -129,4 +129,40 @@ describe('LoopBack Context', function() {
+      runWithPushedValue('test-value-1', true),
+      runWithPushedValue('test-value-2', true),
+    ]).then(function() {
+      return Promise.all([
```
Please create two tests - one using bind, the other not using bind. That way when one of them fails, the second will be still  run.

Both tests can share the same `runWithPushedValue` function, which should be defined *after* the tests - see http://loopback.io/doc/en/contrib/style-guide.html

josieusa replied:
Sorry, I missed this comment before. I'm going to refactor the test accordingly.

josieusa replied:
@bajtos all done

josieusa commented:
Done. I agree, and since I don't think I could ever manage to make the "failure test" robust (even in Node 6), I removed it. We don't need to test further that this bug is reproducible, and anyway that should be done by forking `loopback-sandbox`, not with a test here.

bajtos commented on test/main.test.js:
```diff
@@ -129,4 +129,40 @@ describe('LoopBack Context', function() {
+    ]).then(function() {
+      return Promise.all([
+        runWithPushedValue('test-value-3'),
+        runWithPushedValue('test-value-4', false, true),
```
I think the race condition is caused by the fact that sometimes the first runner fails while other times it's the second runner that fails.

I think we need to rework the tests in such way that the runner only report the value observed in the context, and then the tests asserts that either 1) all values are correct 2) one of the values is not correct, but it doesn't matter which one.

```js
// please use more descriptive test names than "works" and "does not work" :)
it('works', () => {
  return Promise.all([
    runWithPushedValue('test-value-1', {bind: true}),
    runWithPushedValue('test-value-2', {bind: true}),
  ]).then((results) => {
    expect(result).to.eql(['test-value-1', 'test-value-2']);
  });
});

it('does not work', () => {
  return Promise.all([
    runWithPushedValue('test-value-1', {bind: false}),
    runWithPushedValue('test-value-2', {bind: false}),
  ]).then((results) => {
    expect(result).to.not.eql(['test-value-1', 'test-value-2']);
  });
});

function runWithPushedValue(...) {
  // ...
  whenPromise.then(function pullFromContext() {
    var pulledValue = ctx && ctx.get('test-key');
    outerResolve(pulledValue);
  })
  .catch(reject);
}
```

josieusa replied:
I didn't think about it! Anyway, I think that (without bind) one of the two runners always fails, and the other always succeeds. I'm going to refactor the failure test that way, more easily.

josieusa replied:
You were right, sometimes both values are correct. Fixed.

bajtos commented:
@josieusa you pushed a change while I was reviewing the code, please see the four collapsed comments above ☝️  and address those that are still valid.

bajtos commented on README.md:
````diff
@@ -113,6 +112,36 @@ MyModel.myMethod = function(cb) {
   ctx.set('key', { foo: 'bar' });
 });
 ```
+### Bind for concurrency
````
Please add an empty line before the section heading for consistency with the rest of this document.

josieusa committed:
Fix README style

josieusa committed:
Refactor failure test

josieusa committed:
Simplify failure test

josieusa committed:
Account for non-reproducibility of issue

josieusa committed:
Clean up test

josieusa commented:
Cleaned it up, PTAL

josieusa reviewed (commented):
I think that, by using `expect(failureCount).to.be.at.most(1);`, we are hiding under the covers the fact that the attempt to reproduce the bug sometimes succeeds, and sometimes not. So, I personally like it more without the "failure test" (commit d255fc59f78fd7e393156dbcd198293e57cfc7bc ), but that is just personal taste.

josieusa committed:
Separate successful test from failing test

josieusa committed:
Fix typo

josieusa committed:
Experiment to reproduce bug on Travis CI

josieusa committed:
Experiment 2 to reproduce bug on Travis CI

josieusa commented:
Hi, sorry for the spammy commits 😄  I was just experimenting with Travis CI.
By increasing the timeout from 25 to 100 ms, the bug is successfully reproduced on Travis CI, too.
Of course I don't think this is super reliable (in fact it still fails on my machine once every 20-30 times, which is a pretty low rate), but it worked.
So I could do this: `expect(failureCount).to.equal(1);`
I noticed that 100 ms is approximately the time of the first test case on Travis CI, which gives a rough idea of how long do the operations on context take on Travis.
Now this makes more sense; if the operations on context are slow on the platform running the tests, then I suspect that one run could start pulling before some run has finished pushing, for example (i.e. the whole push+pull runs aren't really concurrent).
Maybe I could measure and use that time as the timeout... Or detect if the two runs are really concurrent... I'm not sure how to improve it.
**EDIT** it failed again a commit ago... as I said, not super reliable.

josieusa committed:
Refactor without boolean args in test helper

josieusa committed:
Simplify test using whenPromise.delay()

josieusa commented:
Just one question: is `forEeach` blocking in Node (if you don't do I/O inside the callback)? I've read somewhere that it's blocking. The `getFailureCount()` function in my test relies on the assumption that `forEach` is blocking.

josieusa commented:
I read the spec. Polyfills of forEach (not map) in Javascript code that follow the spec and use process.nextTick (so they are non-blocking) are possible. But I think that all native implementations are blocking. So it's ok to use forEach and return immediately after.

josieusa commented:
Hello,
I think I'm done with this PR.
Comments?

bajtos commented:
Hi @josieusa, I was thinking about this problem for some time now and before I took a closer look at the code again, I have one concern I'd like to discuss with you.

The solution proposed here "fixes" the lost context for the code inside a function that 1) started with the correct context 2) then called context-breaking code. However, it does not fix the lost context for any code that's called after this function.

Here is an example to illustrate what I mean. It uses a middleware, but a similar example can be written using remote hooks, operation hooks or a remote method.

```js
app.use(function(req, res, next) {
  var ctx = LoopBackContext.getCurrentContext({ bind: true });
  doSomethingThatBreaksContext(function(err) {
    if (err) return next(err);
    // cool, ctx.get() still works
    next();
  });
});

app.use(function(req, res, next) {
  var ctx = LoopBackContext.getCurrentContext({ bind: true });
  // ouch, the context broken :(
});
```

I think the solution is to bind all callback invocation to the current context too, i.e.

```js
app.use(function(req, res, next) {
  var ctx = LoopBackContext.getCurrentContext({ bind: true });
  next = ctx.bind(next); // <-- IMPORTANT!
  doSomethingThatBreaksContext(function(err) {
    if (err) return next(err);
    // cool, ctx.get() still works
    next();
  });
});

app.use(function(req, res, next) {
  var ctx = LoopBackContext.getCurrentContext({ bind: true });
  // cool, ctx is good
});
```

Using this principle, perhaps it is better to recommend people to bind all their callbacks, instead of binding the context methods? Using your `when`-based test case, here is what I mean:

```js
        var ctx = LoopBackContext.getCurrentContext();
        expect(ctx).is.an('object');
        ctx.set('test-key', pushedValue);
        var whenPromise = whenV377().delay(timeout);
        whenPromise.then(ctx.bind(function pullFromContextAndReturn() {
          var pulledValue = ctx && ctx.get('test-key');
          outerResolve({
            pulledValue: pulledValue,
            pushedValue: pushedValue,
          });
        })).catch(ctx.bind(reject));
      });
```

That way the new `bind` option is not needed at all.

Another idea to consider in case you prefer to keep the new `bind` option: perhaps  `getCurrentContext({ bind: true })` should bind the `bind` method too, to allow users to fix the CLS context when finishing via the supplied callback?

```js
app.use(function(req, res, next) {
  var ctx = LoopBackContext.getCurrentContext({ bind: true });
  doSomethingThatBreaksContext(function(err) {
    if (err) return ctx.bind(next)(err);
    // cool, ctx.get() still works
    ctx.bind(next)();
  });
});
```

Thoughts?

josieusa commented:
> Another idea to consider in case you prefer to keep the new bind option: perhaps getCurrentContext({ bind: true }) should bind the bind method too, to allow users to fix the CLS context when finishing via the supplied callback?

I prefer this, if it gives you the freedom to call `ctx.bind` wherever you want, even after calling context-breaking code. This way, you also avoid reassigning a function argument, contrary to your example:
`next = ctx.bind(next); // <-- IMPORTANT!`
But I'll think about it, maybe there's an even easier way (or maybe not...)

bajtos commented:
> > Another idea to consider in case you prefer to keep the new bind option: perhaps `getCurrentContext({ bind: true })` should bind the bind method too, to allow users to fix the CLS context when finishing via the supplied callback?

> I prefer this, if it gives you the freedom to call `ctx.bind` wherever you want, even after calling context-breaking code. This way, you also avoid reassigning a function argument

Sounds good. Would you mind extending this pull request to bind `ctx.bind` too, including a unit-test and update of README?

josieusa commented:
sure!

josieusa commented:
Hello,
I tried binding `next` as in your example and it didn't work.
So I thought about a completely different solution.
I already have a test implementation and it works.
What if, instead of passing a boolean as the `bind` option value, we passed the request itself (or the response)? Like this:

    var ctx = LoopBackContext.getCurrentContext({ bind: req });

That way, the first time that `getCurrentContext()` is called (i.e. in the first middleware in your example), its internal implementation could memoize the newly created, "good" context as a value in an ES6 WeakMap (using the passed request as the key for that value).

The second time that `getCurrentContext()`is called (i.e. in the second middleware in your example), by internally using the request as the WeakMap key again, it could return the memoized value, i.e. the same context as before.
The second time it avoids creating a new context altogether, and the risk of creating a "bad" context.

It's important to note that, in this example, this context instance is returned only for this request, and not for other concurrent requests. There is no mixing up.

Problems:
~1.~
~WeakMap is not supported in Node 4. We could use a Symbol or a library that generates a random enough UID.~

~Then, in a middleware that must run before the others, we save the ID (or the Symbol) inside the request, like this:~

~`app.use(function(req, res, next) {`~
~`  var id = createId();`~
~`  req.id = id;`~
~`});`~

~Or maybe `res.locals.id = id;`, which is more idiomatic Express code (in this case you'd need to pass the response to `getCurrentContext`, and not the request).~

~Finally, inside the `getCurrentContext()` implementation, we use that ID (or Symbol) as the map key, but the map is a plain object, no longer a WeakMap.~

~It should delete the context from the map when the response is done.~

**EDIT** maybe WeakMap is supported in Node 4, see http://node.green/

2.
In `runInContext` there is no request. Neither in operation hooks. So you'd need to (generate and) pass an ID (or Symbol) to `getCurrentContext`, instead of the request.

The developer should delete the context from the map when the task is done.

**EDIT** Another way, which would be easier to use, would be to attach a "good" context instance to `res.locals` at the beginning of the middleware chain. Then you could get the context from the `res` object. But I don't like this option, since not all the routes need a context. If the user wishes to do so anyway, he/she could do it himself/herself in a user-defined middleware.

**EDIT 2**
This idea is not very different from using directly `res.locals` to store the values that you would otherwise store in the context. But `res.locals` is not available in `runInContext()`.

josieusa commented:
Regarding the solution that uses a WeakMap, in the case of operation hooks, since `context.hookState` is preserved across "before" and "after" hooks, we could use that object as the key for the WeakMap, couldn't we?
That way, you would write:

    MyModel.observe('before save', (context, next) => {
      var clsContext = LoopbackContext.getCurrentContext({bind: context});
      ...
    }

Why not use `hookState` directly? Mostly for legacy code.

bajtos commented:
> I tried binding next as in your example and it didn't work.

Weird. Are you referring to my code snippets from https://github.com/strongloop/loopback-context/pull/19#issuecomment-273186770? They don't show the implementation in `getCurrentContext`. If simple `boundContext.bind = ns.bind(ns.bind)` does not work, then how about the following?

```js
boundContext.bind = ns.bind(function(fn, context) {
  return ns.bind.apply(ns, arguments);
});
```
> The first time that getCurrentContext() is called (i.e. in the first middleware in your example), its internal implementation could memoize the newly created, "good" context as a value in an ES6 WeakMap (using the passed request as the key for that value).
>
> The second time that getCurrentContext()is called (i.e. in the second middleware in your example), by internally using the request as the WeakMap key again, it could return the memoized value, i.e. the same context as before.
The second time it avoids creating a new context altogether, and the risk of creating a "bad" context.

I am afraid there is no such thing like a good or a bad context instance, at least as far as my understanding of CLS goes. The `ns` object (the context) is a constant wrapper that resolves CLS only at the time when get/set/bind methods are called. So if you attach a "good" context to `req`, and then look it up when the CLS chain is broken, the "good" context will actually become a bad one.

Also code that has access to `req` can store any extra per-request data on the `req` object directly and doesn't need CLS.

TBH, I feel like we are over-engineering this. CLS will never work 100% reliably until it's fixed in Node.js core. We need to draw a line where adding extra reliability is not worth the required increase in code/usage complexity.

In that light, I am proposing to land this patch as it is now, because it provides a nice small improvement that's worth it.

Thoughts?

josieusa commented:
> If simple boundContext.bind = ns.bind(ns.bind) does not work

That's exactly the case. At least for me. I'll try your new example instead, and let you know.

> code that has access to req can store any extra per-request data on the req object directly and doesn't need CLS

Yes, either in `req` or in `res.locals`. But operation hooks don't have direct access to the request, and legacy code can still use this package there. Same for `runInContext()`.

> I am afraid there is no such thing like a good or a bad context instance

When I said, "good instance", I meant that the issue was fixed by re-using the same instance. Of course that doesn't mean that the instance can't get broken in different ways; in fact it could (I didn't test the open connection pool issues yet, for example). The practical answer is to write tests on a case-by-case basis.

> In that light, I am proposing to land this patch as it is now, because it provides a nice small improvement that's worth it.

If we don't patch the "two middlewares" use case in your example, I feel that something is missing. At least some documentation about that case. I'll take a few days to keep thinking about it, while I try your new example.

bajtos commented:
> > In that light, I am proposing to land this patch as it is now, because it provides a nice small improvement that's worth it.

> If we don't patch the "two middlewares" use case in your example, I feel that something is missing. At least some documentation about that case. I'll take a few days to keep thinking about it, while I try your new example.

Sounds good. Take your time and let me know when you come up with something.

josieusa commented:
Hello,
at last, I had the time to test your advice, and it seems to work!
```
boundContext.bind = ns.bind(function(fn, context) {
  return ns.bind.apply(ns, arguments);
});
```
I'll take my time to understand the **magic**, make more tests, and then merge.

josieusa commented:
Hello,
I think I understood the magic in your snippet, more or less.
In fact, this seems to work too:
```
ns.bind.bind(ns);
```
😎😎😎😎😎😎😎

josieusa commented:
Hello again,
I think I understood the magic in your snippet.
In fact, this seems to work too:
```
// Call to Function.prototype.bind(), not ns.bind()
boundContext.bind = ns.bind.bind(ns);
boundContext.get = boundContext.bind(ns.get);
boundContext.set = boundContext.bind(ns.set);
```
😎😎😎😎😎😎😎😎
EDIT I'll write a test case for the "2 middlewares" use case and then merge
EDIT 2: in order to make it work, you'll need this line at the beginning of the "bad" middleware:
`next = context.bind(next);`

josieusa committed:
Fix mixed up ctx in middlewares after a bad one

josieusa committed:
Add test for bad middleware chain use case

josieusa commented:
@bajtos PTAL
Do you think that the test needs to be made more readable, for example?
Thanks a lot!

josieusa committed:
Document case of mixed up ctx in middleware chains

josieusa committed:
Improve README

josieusa committed:
Move middleware example into usage section

josieusa committed:
Explain reason for bug fix

josieusa commented:
This is an alternative snippet to `next = ctx.bind(next);`:

https://github.com/othiym23/node-continuation-local-storage/issues/29#issuecomment-226965698

The idea is the same. It uses `Namespace.prototype.bind()`, too.
But it patches the middleware function object, instead of the function body.
I think the PR is quite ok as it is now.

bajtos committed:
minor improvements

 - fix jsdoc for getCurrentContext
 - clean up getCurrentContext code
 - clean up the new test

bajtos commented:
@josieusa thank you for the update. I did few minor imrovements directly myself - see 442d77f. I you are happy with the current patch, then please squash the commits and we can land this pull request.
